############################# WORKFLOW #############################
# 
# Gamma-Delta_workflow of the Metagenomics workflow will perform mapping of environmental samples
# 1) The reads are filtered and mapped against the reference mitochondrial genome and ribosomal repeats generated by Assmbly method from Genome Skimming
#
###################################################################
#
############################# RULES - PROGRAMS #############################
include:
    '../rules/PreprocessingStand.smk'
include:
    '../rules/Bwa-mem2_Indexing.smk'
include:
    '../rules/RemovedNonMapped.smk'
include:
    '../rules/Gamma-Delta_Algorithm.smk'

############################# RULE - RESULTING ANALYSIS #############################
# This functions as an "rule all" where the input is expected to be the finalized output. 
# By using touch can create an empty file to fake the output. 
rule Analysis:
    input:
        expand(
                [
                os.path.join(DATA_DIR_MG, "{PROJECT}/02_FastQC_Results/{SAMPLE}_R1_Paired_fastqc.html"),
                os.path.join(DATA_DIR_MG, "{PROJECT}/02_FastQC_Results/{SAMPLE}_R1_Unpaired_fastqc.html"),
                os.path.join(DATA_DIR_MG, "{PROJECT}/02_FastQC_Results/{SAMPLE}_R2_Paired_fastqc.html"),
                os.path.join(DATA_DIR_MG, "{PROJECT}/02_FastQC_Results/{SAMPLE}_R2_Unpaired_fastqc.html"),                    
                os.path.join(DATA_DIR_MG, "{PROJECT}/09_Gamma_Delta_Results/{SAMPLE}_Gamma_Delta.csv")
                    
                ],
                PROJECT = config["metagenomics"]["project"],
                SAMPLE = config["metagenomics"]["sample"]                
            )
    output:
        touch('Gamma-Delta_Analysis.done')